src/EventStore.Core/Services/Transport/Tcp/ClientTcpDispatcher.cs

```diff
@@ -300,19 +302,19 @@ private static ClientMessage.ReadEvent
 
 		private ClientMessage.UnsubscribeFromStream UnwrapUnsubscribeFromStream(TcpPackage package, IEnvelope envelope,
 			ClaimsPrincipal user) {
-			var dto = package.Data.Deserialize<TcpClientMessageDto.UnsubscribeFromStream>();
+			var dto = package.Data.Deserialize<UnsubscribeFromStream>();
 			if (dto == null) return null;
 			return new ClientMessage.UnsubscribeFromStream(Guid.NewGuid(), package.CorrelationId, envelope, user);
 		}
 
 		private TcpPackage WrapSubscribedToStream(ClientMessage.SubscriptionConfirmation msg) {
-			var dto = new TcpClientMessageDto.SubscriptionConfirmation(msg.LastIndexedPosition, msg.LastEventNumber);
+			var dto = new SubscriptionConfirmation(msg.LastIndexedPosition, msg.LastEventNumber ?? 0);
```

Comment:
This is unfortunate for the JVM tcp client: `msg.LastEventNumber ?? 0` - Using proto3 the client has no way to differentiate 0 from a valid event number or if it is missing.